Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -62,78 +164,188 @@ public void Initialize()
-                            {
-                                this.logger.LogTrace("Rewound to '{0}', which is still not a part of the current best chain, rewinding further.", hash);
-                                continue;
-                            }
-
-                            this.logger.LogTrace("Rewound to '{0}'.", hash);
-                            this.Tip = rewinded;
-                            this.Puller.SetLocation(rewinded);
-                            break;
+                            this.logger.LogTrace("Rewound to '{0}', which is still not a part of the current best chain, rewinding further.",
+                                hash);
+                            continue;
                         }
+
+                        this.Tip = rewinded;
+                        this.Puller.SetLocation(rewinded);
+                        this.logger.LogInformation("Reorg detected, rewinding from '{0}' to '{1}'.", lastTip, this.Tip);
+                        return;
                     }
                 }
 
-                this.AcceptBlock(new ContextInformation(result, this.Validator.ConsensusParams));
-            }
-            catch (ConsensusErrorException ex)
-            {
-                result.Error = ex.ConsensusError;
+                try
+                {
+                    this.ValidateBlock(new ContextInformation(item, this.Validator.ConsensusParams));
+                }
+                catch (ConsensusErrorException ex)
+                {
+                    item.Error = ex.ConsensusError;
+                }
+
+                if (item.Error != null)
+                {
+                    this.logger.LogError("Block rejected: {0}", item.Error.Message);
+
+                    // Pull again.
+                    this.Puller.SetLocation(this.Tip);
+
+                    if (item.Error == ConsensusErrors.BadWitnessNonceSize)
+                    {
+                        this.logger.LogInformation("You probably need witness information, activating witness requirement for peers.");
+                        this.connectionManager.AddDiscoveredNodesRequirement(NodeServices.NODE_WITNESS);
+                        this.Puller.RequestOptions(TransactionOptions.Witness);
+                        return;
+                    }
+
+                    // Set the chain back to ConsensusLoop.Tip.
+                    this.Chain.SetTip(this.Tip);
+                    this.logger.LogTrace("Chain reverted back to block '{0}' ", this.Tip);
+
+                    // Since ChainHeadersBehavior check PoW, MarkBlockInvalid can't be spammed.
+                    this.logger.LogError("Marking block as invalid.");
+                    this.chainState.MarkBlockInvalid(item.Block?.GetHash());
+                }
+                else
+                {
+                    this.logger.LogTrace("Block accepted '{0}' ", this.Tip);
+
+                    this.chainState.HighestValidatedPoW = this.Tip;
```

Comment:
in the error branch above, how is that we don't set this HVPoW?

Reply:
Yes good point fixed

Reply:
this is done